### ISSUE TITLE
実験: 競りPPOのreward creditをNapoleon契約結果に限定してPASS collapseを再検証

### ISSUE THREAD

**Issue by eletim** [eletim/napoleon-web#362]
Related: #202, #359, #360
Base branch: `feature/issue-202-non-playing-ai`

## 背景

Issue #359 の調査で、現在の bidding PPO では PASS と bid に返る学習信号が大きく非対称になっていることが分かった。

主な結果:

- PASS sample は Citizen / Adjutant 側へ回った結果の相対rewardを受けやすい
- BID sample は Napoleon / Napoleon+Adjutant になり、失敗時の大きな負rewardを受けやすい
- iter99 では PASS sample の 77.4% が Citizen / Adjutant alliance outcome
- BID sample の 94.4% が failed Napoleon / Napoleon+Adjutant outcome
- mean BID advantage は -3.87
- PASS は bootstrap 時点でほぼtop-1のnear-tieだったが、PPO updateでその差が急速に拡大した

Issue #360 では opponent mix を Conservative / All-Pass 50:50 に変更したが、PASS collapse は改善せず、むしろ30 iterationで以下となった。

- PASS rate: 98.9%
- PASS top-1: 100%
- argmax bid: 0%
- candidate Napoleon rate: 1.2%
- All Pass: 78.9%
- PASS-bestBid gap: 5.27

したがって、次は opponent distribution ではなく bidding decision に返す reward / credit assignment を最小変更して検証する。

## 問題

現在は candidate の bidding decision に、その後 candidate が Citizen / Adjutant として得た終局rewardまで返っている。

しかし競りActorの観点では、

- candidate がPASSした
- 他プレイヤーがNapoleonになった
- そのNapoleonが失敗した
- candidate がCitizen / Adjutantとして勝った

という結果を、PASSそのものの成功として強く学習させるのは不自然である。

競りActorが本来学習すべき中心は、candidate自身が契約を取りに行った場合の成否と、契約を取らなかった場合のneutralな判断である。

## 目的

bidding PPO専用の credit assignment を導入し、candidate が Napoleon 契約を取らなかったゲームから Citizen / Adjutant の正rewardを bidding decisionへ返さないようにする。

PPO / network / observation / opponent mix / playing policy は変更せず、reward attributionだけの効果を切り分ける。

## Bidding-specific reward v1

bidding training sample に返す reward を、candidate自身の最終的な契約結果に基づいて定義する。

### candidate が Napoleon または Napoleon+Adjutant になった場合

現行 Reward v3 の candidate本人の raw terminal reward を使用する。

宣言枚数を `d` とする。

- Napoleon win: `2 * d`
- Napoleon lose: `-5`
- Napoleon+Adjutant win: `3 * d`
- Napoleon+Adjutant lose: `-5`

### candidate が Napoleon 契約を取らなかった場合

以下は bidding reward を **0** とする。

- Citizen
- Adjutant
- all-pass starter
- all-pass other
- candidate が途中でbidしたが最終的に別playerが契約を取った場合

つまり、他playerのNapoleon成功/失敗によってcandidateのbidding rewardを正負に動かさない。

## relative reward transform

この bidding-specific reward には、現在の

`raw_reward_i - mean(raw_reward_all_players)`

を適用しない。

理由:

- 他playerのNapoleon失敗を平均との差し引きすると、candidateがCitizen/Adjutantであるだけで正rewardになり、今回除去したいcreditが再導入される
- bidding Actorではcandidate自身の契約判断へ直接creditを与えることを優先する

したがって bidding PPO のみ、上記 bidding-specific raw reward をそのままtraining returnとして使う。

choosing-adjutant / exchange のreward設計はこのIssueでは変更しない。

## 重要な意味

PASSそのものに罰則を与えるIssueではない。

弱い手でPASSし、他playerが契約を取った場合は reward=0 とする。

強い手でbidしてNapoleonになり成功すれば大きな正reward、失敗すれば負rewardとなる。

これにより、

- 「他人が失敗したのでPASSして得をした」

という競りActorに不要なsignalだけを除去する。

## 維持する条件

reward credit assignment以外は変えない。

- candidate x1 + Frozen x4 topology
- candidate seat 5席rotation
- Frozen bidding opponent: Issue #360 の Conservative / All-Pass 50:50
- temperature=1 rollout sampling
- compact278 observation / model
- strongest-suit-balanced minibatch
- PPO clip epsilon
- entropy coefficient
- advantage normalization setting
- learning rate / epochs / batch size
- Frozen playing policy `ppo-separated-v1000`
- adjutant / exchange artifact
- game rule
- all-pass game-core処理

## provenance / resume safety

新しい bidding reward / credit assignment を明確なversionとして保存する。

最低限:

- bidding reward id/version
- non-Napoleon candidate reward = 0
- relative transform disabled for bidding
- Napoleon / Napoleon+Adjutant reward values

旧reward runを新設定で誤resumeしないよう fail-close する。

## diagnostics

Issue #359 / #360 で使ったdiagnosticsを維持する。

最低限各checkpointで:

- PASS rate / bid rate
- PASS top-1 rate
- argmax bid rate
- PASS probability
- best legal bid probability
- PASS-bestBid gap
- policy entropy
- candidate Napoleon formation rate
- declaration success rate
- All Pass rate
- candidate role distribution
- PASS vs BID sample count
- PASS vs BID mean/median return
- PASS vs BID mean/median advantage
- positive advantage rate

さらに role別に:

- Napoleon
- Napoleon+Adjutant
- Citizen
- Adjutant
- All Pass

の sample count / training reward / advantage を確認し、Citizen/Adjutantが bidding reward 0 になっていることを検証する。

## 短期再学習

まず30 iterationで確認する。

Issue #360 と可能な限り同条件を使う。

推奨:

```text
iterations             30
games-per-iteration    200 logical seeds
epochs                  4
batch-size              128
learning-rate           1e-4
bidding entropy coef    0.1
advantage normalization none
minibatch strategy      strongest-suit-balanced
seed                    202
temperature             1
```

inference device は利用可能なbackendを使ってよいが、比較条件として記録する。
training device はCUDA利用可能ならCUDAを使う。

旧runは上書きしない。

## 比較対象

主比較:

### Issue #360

`non-playing-ppo-compact278-bidding-only-conservative-allpass50-e01-lr1e4-30x200-cpu`

iter30:

- PASS rate 98.9%
- PASS top-1 100%
- argmax bid 0%
- PASS prob 96.7%
- best bid prob 0.5%
- PASS-bestBid gap 5.27
- Napoleon rate 1.2%
- All Pass 78.9%

### current compact278

RuleBased / Conservative 50:50 の既存100 iteration runも参考値として残す。

## 成功判定

単にbid率が増えれば成功とはしない。

望ましい方向:

- PASS top-1が100%張り付きから外れる
- argmax bidが0%から明確に増える
- 強いhandでbid top-1が発生する
- candidate Napoleon rateが回復する
- PASS / BID advantage差が縮小する
- Citizen/Adjutant outcomeがPASS advantageを押し上げなくなる
- stochastic PASS率が0%近傍へ逆collapseしない
- Napoleon成功率が極端に低いまま無条件bidへ向かわない

## Hand-strength diagnostics

最低限、RuleBased hand-strength bin別に以下を見る。

- sample count
- selected PASS rate
- PASS top-1 rate
- argmax bid rate
- PASS probability
- best bid probability
- PASS-bestBid gap
- mean return / advantage

特に強いhandでbid top-1が現れるかを確認する。

## 判定したいこと

最終報告では以下へ明示的に回答する。

1. Citizen/Adjutant rewardをbidding creditから除外するとPASS collapseは改善するか
2. PASS top-1 100%は解消するか
3. argmaxでbidするhandが出現するか
4. 強いhandほどbidする傾向が現れるか
5. PASS vs BID advantage差はどの程度縮まるか
6. candidate Napoleon率 / declaration success / All Pass率はどう変化するか
7. reward attributionだけで十分か
8. 次に必要なのは reward magnitude調整 / action balancing / PPO設定 / observation conditioning のどれか

## Git / PR運用

このIssueはreward attributionの実装変更を含むため、最新 `feature/issue-202-non-playing-ai` からIssue専用ブランチを作成する。

PR baseは `feature/issue-202-non-playing-ai` とする。

実装・unit test・smoke・30 iteration短期run完了後、Codex review最大4回（review後修正最大3回）。

blocking / actionable 指摘がなく必要テストが成功したら Ready for review にして親ブランチへmergeする。

`main`にはmergeしない。

結果はIssueへコメントし、必要なら `experiments/` 配下にも記録する。

## Acceptance Criteria

- bidding-specific reward v1 が導入される
- candidateがNapoleon/Napoleon+Adjutantの場合のみ現行Reward v3相当の本人rewardをbiddingへ返す
- Citizen / Adjutant / All Pass / auction lost は bidding reward 0
- biddingではgame-mean relative transformを適用しない
- choosing-adjutant / exchange rewardは変更しない
- reward version / provenanceが保存される
- 旧runとのresume safetyがある
- unit testでrole/outcome別rewardを固定fixture確認する
- smokeで fallback=0 / illegal action=0
- 30 iteration短期runを実施する
- PASS top-1 / argmax bid / PASS-bestBid gap / PASS-vs-BID advantageを比較する
- hand-strength別にargmax bid発生を確認する
- Issue #360と具体的な数値比較をIssueへ残す

## 非目標

- PPO hyperparameter変更
- opponent mix変更
- network architecture変更
- observation schema変更
- playing policy再学習
- choosing-adjutant / exchange reward変更
- game rule変更
- mainへのMerge
